src/jit/expr.c

```diff
@@ -0,0 +1,1006 @@
+        MVMuint16 opcode = ins->info->opcode;
+        MVMSpeshAnn *ann;
+        const MVMJitExprTemplate *template;
+        MVMint32 before_label = -1, after_label = -1,
+            wrap_before = 0, wrap_after = 0;
+        struct ValueDefinition *defined_value = NULL;
+        if (opcode == MVM_SSA_PHI || opcode == MVM_OP_no_op) {
+            continue;
+        }
+
+        /* check if this is a getlex and if we can handle it */
+        BAIL(opcode == MVM_OP_getlex && !can_getlex(tc, jg, ins), "Can't compile object getlex");
+
+        /* Check annotations that may require handling or wrapping the expression */
+        for (ann = ins->annotations; ann != NULL; ann = ann->next) {
+            MVMint32 idx;
+            switch (ann->type) {
+            case MVM_SPESH_ANN_FH_START:
+                /* start of a frame handler (inclusive). We need to mark this
+                 * instruction with a label so that we know the handler covers
+                 * this code, and we need to install a dynamic label to tell the
+                 * interpreter that JIT execution has reached this point. */
+                before_label = MVM_jit_label_before_ins(tc, jg, iter->bb, ins);
+                jg->handlers[ann->data.frame_handler_index].start_label = before_label;
+
+                wrap_before = MVM_JIT_CONTROL_DYNAMIC_LABEL;
+                break;
+            case MVM_SPESH_ANN_FH_END:
+                /* end of the frame handler (exclusive), funnily enough not the
+                 * end of a basic block. The dynamic label will be installed
+                 * just after the label that marks the end, which means that
+                 * jit_entry_label will compare greater-than the end label,
+                 * which ensures that the end is exclusive. */
+                before_label = MVM_jit_label_before_ins(tc, jg, iter->bb, ins);
+                jg->handlers[ann->data.frame_handler_index].end_label = before_label;
+                wrap_before = MVM_JIT_CONTROL_DYNAMIC_LABEL;
+                break;
+            case MVM_SPESH_ANN_FH_GOTO:
+                /* A label to jump to for when a handler catches an
+                 * exception. Thus, this can be a control flow entry point (and
+                 * should be the start of a basic block, but I'm not sure if it
+                 * always is).  */
+                before_label = MVM_jit_label_before_ins(tc, jg, iter->bb, ins);
+                jg->handlers[ann->data.frame_handler_index].goto_label = before_label;
+                active_values_flush(tc, tree, values, sg->num_locals);
+                break;
+            case MVM_SPESH_ANN_DEOPT_OSR:
+                /* A label the OSR can jump into to 'start running', so to
+                 * speak. As it breaks the basic-block assumption, arguably,
+                 * this should only ever be at the start of a basic block. But
+                 * it's not. So we have to insert the label and compute it. */
```

Comment:
As of a while ago, this actually is the case: https://github.com/MoarVM/MoarVM/blob/master/src/spesh/graph.c#L329

Reply:
If that's the case then this can be cleaned up. Either way no label will actually be inserted since `MVM_jit_is_label_for_ins` will be false (it is a basic block label in that case) 😄